ZT crashing when viewing a message in the browser.

Sushmey opened the pull request:
<!-- See README for documentation, or ask in #zulip-terminal if unclear -->
### What does this PR do, and why?

ZT crashes on MacOS when opening a message in browser due to an attribute error, where it tries to get 'name' to display the browser name. This might be an issue for other OS too.

Also refactored the test function name, since it had an extra '_'
### External discussion & connections
<!-- [x] all that apply, specifying topic and adding numbers after # for issues/PRs -->
- [x] Discussed in **#zulip-terminal** in `ZT crashes trying to view message in browser #T1471 #T1472`
- [x] Fully fixes #1471 
- [ ] Partially fixes issue #
- [ ] Builds upon previous unmerged work in PR #
- [ ] Is a follow-up to work in PR #
- [ ] Requires merge of PR #
- [ ] Merge will enable work on #

### How did you test this?
<!-- [x] all that apply -->
- [x] Manually - Behavioral changes
- [x] Manually - Visual changes
- [ ] Adapting existing automated tests
- [ ] Adding automated tests for new behavior (or missing tests)
- [x] Existing automated tests should already cover this (*only a refactor of tested code*)

### Self-review checklist for each commit
- [x] It is a [minimal coherent idea](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development)
- [x] It has a commit summary following the [documented style](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development) (title & body)
- [x] It has a commit summary describing the  motivation and reasoning for the change
- [x] It individually passes linting and tests
- [ ] It contains test additions for any new behavior
- [ ] It flows clearly from a previous branch commit, and/or prepares for the next commit

### Visual changes    <!-- DELETE SECTION IF NO VISUAL CHANGE -->
<!-- Zulip tips at https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html -->
<!-- For video, try asciinema; after uploading, embed using
[![yourtitle](https://asciinema.org/a/<id>.png)](https://asciinema.org/a/<id>)
-->
<!--NOTE: Attached videos/images will be clearer from smaller terminal windows -->

<img width="1440" alt="Screenshot 2024-02-15 at 10 13 09 PM" src="https://github.com/zulip/zulip-terminal/assets/71403193/42884f19-4f1f-477b-9c59-10e08edd9d16">

neiljp reviewed (commented):
@Sushmey Thanks for reporting and proposing a solution for this. I can't test this locally, but this fix looks like it would work :+1:

I gave some feedback in the stream, including re the test name change being unnecessary, and if we can improve the comment/commit-text.

Also, while this is a quick fix, some small notes on the solution:
- setting the attribute to itself is odd to read
- I'd typically extract the name into a new variable, instead of setting an attribute on an object
- `getattr` may be a cleaner solution, which also can provide a default
- the `.` in the exceptional case seems inconsistent?
- adding a test (or test case) for this issue would be useful, and might illustrate the `.` difference too ;)

Sushmey commented:
Hi Neil, thanks for the review!
I made the necessary changes

- Changed the commit text to include PY version discrepancy. 
- Used `getattr` with a default value.
- Removed the period all together.
- Added a test case.
- Changed the name of the test back to original name.

Sushmey commented:
@zulipbot remove "PR awaiting update" 
add "PR needs review"

neiljp reviewed (commented):
@Sushmey Thanks for the update and looking at a test in more detail :+1:

The inline comments are primarily regarding the test, since it's currently not quite testing what you expect, and structuring the input/output parameters to the test will help make things clearer.

Sushmey commented on tests/core/test_core.py:
```diff
@@ -406,20 +406,26 @@ def test_copy_to_clipboard_exception(
 
     @pytest.mark.parametrize(
-        "url",
+        "url, name, name_attribute_present",
```
Are you suggesting to change `name` to `expected_name` with the type Optional[str]? And remove the `name_attribute_present` alltogether? But that would mean the attribute exists and has a default value of "None" right?

neiljp replied:
So we're trying to test the result of `webbrowser.get()` having a name attribute; for the test, we want to have
- 'input' element(s) in the parametrize which represent whether it has the name attribute and if it does, what the value is => how to set up the state before calling the function
- an 'output' element in the parametrize which represents what we're going to assert on.

The first sentence of my previous comment was related to how `name_attribute_present` was one aspect of the 'input'. The second paragraph was explaining how that can be represented, which you've updated for.

My third paragraph was indicating that we keep/add another parameter for the 'output', with the `expected` prefix, which would be the value we assert on.

Sushmey replied:
This makes a lot of sense, thanks! I've added `expected_attribute_present`  as the input element, and `webbrowser_name` as the output.

Sushmey commented on tests/core/test_core.py:
```diff
@@ -406,20 +406,26 @@ def test_copy_to_clipboard_exception(
         os.environ["DISPLAY"] = ":0"
         mocked_report_success = mocker.patch(MODULE + ".Controller.report_success")
         mock_get = mocker.patch(MODULE + ".webbrowser.get")
+        mock_get.return_value.name = name
```
Changing it so that I'm not explicitly setting the `mock_get.return_value.name` at all and adding a condition to check if `expected_name` is false (since None is a falsy value) and setting the `mock_get.return_value.name` as "default browser".

This way we can illustrate the behaviour change and the existing test passes even without changing anything else.

neiljp replied:
The point I was making is that we want the name attribute to be conditionally *absent*, representing the macos case for python <3.11.

The value we assign here is what `mock_get`, ie. `webbrowser.get()`, will have as a return value - so hopefully you can see that setting it to be `"default browser"` in the conditional doesn't represent the situation to test?

Sushmey commented on tests/core/test_core.py:
```diff
@@ -429,6 +435,11 @@ def test_open_in_browser_success(
             [f"The link was successfully opened using {mock_get.return_value.name}"]
         )
 
+        if name_attribute_present:
+            assert mock_get.return_value.name != "default browser"
+        elif not name_attribute_present:
+            assert mock_get.return_value.name == "default browser"
+
```
Yeah I did sense that this seemed wrong while writing it but thought I was being thorough. Anyway, removing this whole if...else block. Doesn't seem important.

Sushmey commented on zulipterminal/core.py:
```diff
@@ -414,11 +414,9 @@ def open_in_browser(self, url: str) -> None:
             # Suppress stdout and stderr when opening browser
             with suppress_output():
                 browser_controller.open(url)
+                browser_name = getattr(browser_controller, "name", "default browser")
```
Alright, adding this comment

`MacOS using Python version < 3.11 has no "name" attribute hence setting a default value`

neiljp reviewed (commented):
@Sushmey I've tried to clarify what I meant in my last review.

I've also commented elsewhere on older threads, since they can't be added to this review but the context is useful for a reply.

I suspect you may want to read more about mocking. The mocker items are wrapped versions from unittest.mock, so that's a good reference to use.

Another useful aspect, is to adjust the code once you think your tests should check it works properly, and see what happens. Another tactic is to make no code changes and see if it accidentally passes, which one can do by having the tests and code (initially) in different commits and switching their order, and running tests after only the test commit.

Sushmey commented:
All these comments are really helpful making me understand how to write better tests. I've made changes incorporating all your feedback. The test changes are as follows, 

There are two new parameters, `webbrowser_name` and `expected_attribute_present`, the first one is what the name attribute is supposed to be if it has one, and the second one tells if the test case has name attribute or not.

Then we check if `webbrowser_name` is None, if yes, then delete the name attribute from the mock since that's the case we are testing for. If `webbrowser_name` exists, then we just set the name attribute of the mock as `webbrowser_name`.

Finally, we check if the mock has a success report with the success message with `mock_get.return_value.name` if `expected_attribute_present` is true (i.e. name attribute does exist for this case) or if the attribute doesn't exist, if the success message has `"default browser"` instead of the web browser name.

neiljp commented:
@Sushmey The test in this iteration looks like it should validate the code properly now - did you add the part I suggested, and then add the last conditional to make it pass?

The final point I was getting at was that the last parameter can simply be a string which represents the expected output value, *which you can use in the assert*. This is why I previously suggested/hinted the name of that parameter might not be good to keep as `expected_attribute_present`. 

While using the `mock_get` value in the assert looks like it works now, before accepting code that asserts on the same mock values that are assigned in a test, I'd aim to examine it very closely for validity. Here I suspect it couldn't work without that extra conditional at the end.

By using a string for that parametrize value, you can remove that extra conditional (`if`) very easily. It does lead to slight duplication of strings, but makes the test case data more explicit when you read the top; this currently reads as something like
```python
            "url, webbrowser_name, expected_attribute_present",
            ...
            ("https://chat.zulip.org/user_uploads/sent/abcd/efg.png", "mozilla", True),
            ("https://github.com/", None, False),
```
To understand the values of `expected_attribute_present` just based on the above, that parameter must be named very clearly, or else the test function must be read and understood.

When it's a string instead, which in some cases matches the input string, then it can be clearer what may be happening before you read the rest:
```python
            "url, webbrowser_name, expected_<descriptive name for output>",
            ...
            ("https://chat.zulip.org/user_uploads/sent/abcd/efg.png", "mozilla", "mozilla"),
            ("https://github.com/", None, "default browser"),
```

Sushmey commented:
The last conditional was to validate the success message and the conditional seemed necessary since I wasn't sure if it's worth the trade-off for adding more parameters, albeit slightly redundant, just so that we could get rid of the conditionals in the test. I've added it now. Going forward, I'll try to reduce conditionals if adding parameters helps reduce them since it also makes the tests easier to understand at a glance. 

Added a new parameter of type `str` called `expected_webbrowser_name`

neiljp commented:
@Sushmey Thanks for your work on this :+1: I just pushed here with some textual adjustments and will now merge :tada:
